app/controllers/products_controller.rb

```diff
@@ -0,0 +1,171 @@
+  before_action :find_product, only: [:show, :edit, :update, :destroy, :retired]
+  before_action :initialize_search, only: [:index]
+
+  def index
+    @products = Product.where(retired: false)
+    @users = User.all
+    @categories_select = Category.alphabetical
+
+    handle_search_name
+    handle_filters
+  end
+
+  def clear
+    clear_session(:search_name, :filter_name, :filter)
+    redirect_to products_path
+  end
+
+  def show
+    if @product.nil?
+      render_404
+      return
+    end
+  end
+
+  def new
+    if session[:user_id].nil?
+      flash[:error] = "You cannot add a new product without signing in."
+      redirect_to root_path
+      return
+    end
+    @product = Product.new
+  end
+
+  def create
+    @product = Product.new(product_params)
+    @product.user_id = session[:user_id]
+    if @product.save
+      flash[:success] = "Product has been successfully added"
+      redirect_to product_path(@product.id)
+    else
+      flash[:error] = "Product has not been added"
+      render :new, status: :bad_request
+      return
+    end
+  end
+
+  def edit
+    if @product.nil?
+      flash[:error] = "Product not found"
+      redirect_to products_path
+      return
+    end
+    if session[:user_id] != @product.user.id
+      flash[:error] = "You cannot edit another user's products."
+      redirect_to root_path
+      return
+    end
+  end
+
+  def update
+    if @product.nil?
+      flash[:error] = "Product not found"
+      redirect_to products_path
+      return
+    elsif @product.update(product_params)
+      flash[:success] = "Product has been successfully updated"
+      redirect_to product_path # go to the product details page
+      return
+    else # save failed
+      flash.now[:error] = "Product has not been updated"
+      render :edit, status: :bad_request # show the new product form view again
+      return
+    end
+  end
+
+  def destroy
+    if @product.nil? 
+      render_404
+      return
+    end
+    
+    if session[:user_id].nil?
+      flash[:error] = "You must be logged in to delete this item"
+      # need to clarify which path to redirect
+      redirect_to products_path
+      return
+    elsif session[:user_id] == @product.user_id
+      @product.update_attribute(:retired, true)
+      @product.destroy
+      flash[:success] = "Successfully destroyed #{ @product.name }"
+      # need to clarify which path to redirect
+      redirect_back(fallback_location: root_path)
+      return
+    else
+      flash[:error] = "Only the product Seller can delete the product."
+      # need to clarify which path to redirect
+      redirect_to root_path
+      return
+    end
+  end
+
+  def retired
+    if @product.nil? 
+      render :file => "#{Rails.root}/public/404.html", layout: false, status: :not_found
+      return
+    end
+
+    if session[:user_id].nil?
+      flash[:error] = "You must be logged in to retire this item"
+      redirect_to root_path
+      return
```

Comment:
This logic is in a couple places. Consider moving it to a controller filter `require_login`.

Reply:
Jessica requested feedback: I would consider adding the `require_login` method to the applications controller because this functionality is used in multiple controllers.